File: ios/Classes/StarprinterPlugin.m
#import "StarprinterPlugin.h"
#import "Printer.h"
#import "Receipt.h"
#import "BarcodeReader.h"

@implementation PrinterReceipt
@synthesize printer;
@synthesize receipt;
-(id) initWithPrinter: (Printer*) printer {
    self = [super init];

    self.printer = printer;
    return self;
}
@end


@implementation StarprinterPlugin
NSDictionary *printerReceipts;
BarcodeScanner *barcodeReader;
FlutterMethodChannel* channel;
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    printerReceipts = [[NSMutableDictionary<NSString*,PrinterReceipt*> alloc]init];
  channel = [FlutterMethodChannel
      methodChannelWithName:@"getzuza.starprinter"
            binaryMessenger:[registrar messenger]];
  StarprinterPlugin* instance = [[StarprinterPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}


- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    
    if ([call.method  isEqual: @"searchPrinters"]) {
        [Printer searchPrinters:^(NSArray<PrinterInfo *> *searchData) {
            
            NSMutableArray *printers = [[NSMutableArray alloc] init];
            
            for (PrinterInfo *printer in searchData) {
                [printers addObject:@{
                    @"modelName": printer.modelName,
                    @"macAddress": printer.macAddress,
                    @"portName": printer.portName
                }];
            }
            result(printers);
        }];
    } else if ([call.method  isEqual: @"getPrinter"]){
        NSString *portName = call.arguments[@"portName"];
        NSNumber *timeOut = call.arguments[@"timeOut"];

        Printer *printer = [Printer getPrinter:portName timeout:[timeOut intValue]];
        PrinterReceipt *printerReceipt = [[PrinterReceipt alloc] initWithPrinter:printer];
        [printerReceipts setValue:printerReceipt forKey:portName];
        result(portName);

    } else if ([call.method  isEqual: @"createReceipt"]){
        NSString *portName = call.arguments[@"portName"];
        bool textValue = call.arguments[@"text"];
        NSNumber *paperSize = call.arguments[@"paperSize"];

        PrinterReceipt *printerReceipt = [printerReceipts objectForKey:portName];
        [printerReceipt.printer createReceiptText:textValue PaperSize:[paperSize intValue] Handler:^(Receipt *receipt) {
            printerReceipt.receipt = receipt;
            result(portName);
        }];

    } else if ([call.method  isEqual: @"addAlignLeft"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addAlignLeft];
        result(nil);

    } else if([call.method  isEqual: @"addAlignRight"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addAlignRight];
        result(nil);

    } else if ([call.method  isEqual: @"addAlignCenter"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addAlignCenter];
        result(nil);

    } else if ([call.method  isEqual: @"setBlackColor"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt setBlackColor];
        result(nil);

    } else if ([call.method  isEqual: @"setRedColor"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt setRedColor];
        result(nil);

    } else if ([call.method  isEqual: @"addText"]){
        NSString *portName = call.arguments[@"portName"];
        NSString *value = call.arguments[@"value"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addText:value];
        result(nil);

    } else if ([call.method  isEqual: @"addDoubleText"]){
        NSString *portName = call.arguments[@"portName"];
        NSString *value = call.arguments[@"value"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addDoubleText:value];
        result(nil);

    } else if ([call.method  isEqual: @"addBoldText"]){
        NSString *portName = call.arguments[@"portName"];
        NSString *value = call.arguments[@"value"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addBoldText:value];
        result(nil);

    } else if ([call.method  isEqual: @"addUnderlinedText"]){
        NSString *portName = call.arguments[@"portName"];
        NSString *value = call.arguments[@"value"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addUnderlinedText:value];
        result(nil);

    }else if ([call.method  isEqual: @"addInverseText"]){
        NSString *portName = call.arguments[@"portName"];
        NSString *value = call.arguments[@"value"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addInverseText:value];
        result(nil);

    } else if ([call.method  isEqual: @"addLine"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addLine];
        result(nil);

    } else if ([call.method  isEqual: @"addImage"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        FlutterStandardTypedData *imgBytes = call.arguments[@"bytes"];
        NSNumber *width = call.arguments[@"width"];
        UIImage *imageData = [UIImage imageWithData:imgBytes.data];
        [receipt.receipt addImage:imageData width:[width intValue]];
        result(nil);

    } else if ([call.method  isEqual: @"addBarcode"]){
        NSString *portName = call.arguments[@"portName"];
        NSString *value = call.arguments[@"value"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        NSNumber *height = call.arguments[@"height"];
        [receipt.receipt addBarcode:value height:[height intValue]];
        result(nil);

    } else if ([call.method  isEqual: @"addQrCode"]){
        NSString *portName = call.arguments[@"portName"];
        NSString *value = call.arguments[@"value"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt addQrCode:value];
        result(nil);
        
    } else if ([call.method  isEqual: @"openCashDrawer"]){
        NSString *portName = call.arguments[@"portName"];
        NSNumber *drawer = call.arguments[@"drawer"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt openCashDrawer:[drawer intValue]];
        result(nil);
        
    } else if ([call.method  isEqual: @"cutPaper"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt cutPaper];
        result(nil);
        
    } else if ([call.method  isEqual: @"closeReceipt"]){
        NSString *portName = call.arguments[@"portName"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt closeReceipt];
        result(nil);
        
    } else if ([call.method  isEqual: @"setCommand"]){
        NSString *portName = call.arguments[@"portName"];
        NSString *command = call.arguments[@"command"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt setCommand:command];
        result(nil);
        
    } else if ([call.method  isEqual: @"setFontSize"]){
        NSString *portName = call.arguments[@"portName"];
        NSNumber *size = call.arguments[@"size"];
        PrinterReceipt *receipt = [printerReceipts objectForKey:portName];
        [receipt.receipt setFontSize: [size doubleValue]];
        result(nil);
        
    } else if ([call.method  isEqual: @"printReceipt"]){
        NSString *portName = call.arguments[@"portName"];
        NSNumber *delay = call.arguments[@"delay"];
        NSNumber *retry = call.arguments[@"retry"];
        NSLog(@"printReceipt called");
        PrinterReceipt *printer = [printerReceipts objectForKey:portName];
        [printer.printer printReceipt:printer.receipt withDelay:[delay doubleValue] andRetry:[retry intValue] onSuccess:^(PrinterStatus *printerStatus) {
            NSDictionary *status = @{
                @"offline": [NSNumber numberWithBool:printerStatus.offline],
                @"error": printerStatus.error,
                @"outOfPaper": [NSNumber numberWithBool: printerStatus.outOfPaper],
                @"paperJam": [NSNumber numberWithBool: printerStatus.paperJam],
                @"printed": [NSNumber numberWithBool: printerStatus.printed]
            };
            result(status);
        } onFail:^(PrinterStatus *printerStatus) {
            NSDictionary *status = @{
                @"offline": [NSNumber numberWithBool:printerStatus.offline],
                @"error": printerStatus.error,
                @"outOfPaper": [NSNumber numberWithBool: printerStatus.outOfPaper],
                @"paperJam": [NSNumber numberWithBool: printerStatus.paperJam],
                @"printed": [NSNumber numberWithBool: printerStatus.printed]
            };
            result(status);
        }];
    } else if ([call.method  isEqual: @"connect"]){
        NSString *portName = call.arguments[@"portName"];
        barcodeReader = [[BarcodeScanner alloc] initWithPortName:portName];
        [barcodeReader setBarcodeScanner:^(NSString *code) {
            NSMutableDictionary<NSString*,NSString*> *data =  [[NSMutableDictionary<NSString*,NSString*> alloc]init];
            [data setValue:code forKey:@"code"];
            [channel invokeMethod:@"onBarcodeRead" arguments:data];
        }];
        result(nil);
    } else if ([call.method  isEqual: @"disconnect"]){
        NSString *portName = call.arguments[@"portName"];
        [barcodeReader disconnect];
        barcodeReader = nil;
        result(nil);
    } else {

    }
}

@end

File: android/src/main/java/com/getzuza/starprinter/StarprinterPlugin.java
package com.getzuza.starprinter;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.HashMap;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;

/** StarprinterPlugin */
public class StarprinterPlugin implements FlutterPlugin, MethodCallHandler, ActivityAware {
  private static final String Class = "StarprinterPlugin";
  class PrinterReceipt {
    public Printer printer;
    public Receipt receipt;

    public PrinterReceipt(Printer printer) {
      this.printer = printer;
    }
  }
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  private static final String STAR_PRINTER = "getzuza.starprinter";
  private HashMap<String, PrinterReceipt> printerReceipts = new HashMap<>();
  private Activity activity;
  BarcodeReader barcodeReader;
  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(),STAR_PRINTER);
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {

    new Handler(Looper.getMainLooper()).post(new Runnable() {
      @Override
      public void run() {
        PrinterReceipt printer;
        String value;
        switch (call.method) {
          case "searchPrinters":
            PrinterInfo[] printers = Printer.searchPrinters(activity);
            ArrayList<HashMap<String, Object>> list = new ArrayList<>();
            for (PrinterInfo p : printers) {
              HashMap<String, Object> hash = new HashMap<>();
              hash.put("portName", p.portName);
              hash.put("modelName", p.modelName);
              hash.put("macAddress", p.macAddress);
              list.add(hash);
            }
            result.success(list);
            break;
          case "getPrinter":
            String portName = call.argument("portName");
            int timeOut;
            if (call.argument("timeOut") != null) {
              timeOut = call.argument("timeOut");
            } else {
              timeOut = 30000;
            }
            printerReceipts.put(portName, new PrinterReceipt(Printer.getPrinter(activity, portName, timeOut)));
            result.success(portName);
            break;
          case "createReceipt":
            portName = call.argument("portName");
            boolean text = call.argument("text");
            int paperSize = call.argument("paperSize");
            printer = printerReceipts.get(portName);
            if (printer != null) {
              printer.receipt = printer.printer.createReceipt(text, paperSize);
              result.success(portName);
            } else {
              result.success(null);
            }
            break;
          case "addAlignLeft":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            printer.receipt.addAlignLeft();
            result.success(null);;
            break;

          case "addAlignRight":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            printer.receipt.addAlignRight();
            result.success(null);;
            break;

          case "addAlignCenter":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            printer.receipt.addAlignCenter();
            result.success(null);;
            break;

          case "setBlackColor":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            printer.receipt.setBlackColor();
            result.success(null);;
            break;

          case "setRedColor":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            printer.receipt.setRedColor();
            result.success(null);;
            break;

          case "addText":
            portName = call.argument("portName");
            value = call.argument("value");
            printer = printerReceipts.get(portName);
            printer.receipt.addText(value);
            result.success(null);;
            break;

          case "addDoubleText":
            portName = call.argument("portName");
            value = call.argument("value");
            printer = printerReceipts.get(portName);
            printer.receipt.addDoubleText(value);
            result.success(null);;
            break;

          case "addBoldText":
            portName = call.argument("portName");
            value = call.argument("value");
            printer = printerReceipts.get(portName);
            printer.receipt.addBoldText(value);
            result.success(null);;
            break;

          case "addUnderlinedText":
            portName = call.argument("portName");
            value = call.argument("value");
            printer = printerReceipts.get(portName);
            printer.receipt.addUnderlinedText(value);
            result.success(null);;
            break;

          case "addInverseText":
            portName = call.argument("portName");
            value = call.argument("value");
            printer = printerReceipts.get(portName);
            printer.receipt.addInverseText(value);
            result.success(null);;
            break;
          case "addLine":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            printer.receipt.addLine();
            result.success(null);;
            break;
          case "addImage":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            byte[] bytes = call.argument("bytes");
            Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0,
                    bytes.length);
            int width = call.argument("width");
            printer.receipt.addImage(bitmap,width);
            result.success(null);;
            break;
          case "addBarcode":
            portName = call.argument("portName");
            value = call.argument("value");
            int height = call.argument("height");
            printer = printerReceipts.get(portName);
            printer.receipt.addBarcode(value,height);
            result.success(null);
            break;
          case "addQrCode":
            portName = call.argument("portName");
            value = call.argument("value");
            printer = printerReceipts.get(portName);
            printer.receipt.addQrCode(value);
            result.success(null);
            break;
          case "openCashDrawer":
            portName = call.argument("portName");
            int drawer = call.argument("drawer");
            printer = printerReceipts.get(portName);
            printer.receipt.openCashDrawer(drawer);
            result.success(null);;
            break;
          case "cutPaper":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            printer.receipt.cutPaper();
            result.success(null);;
            break;
          case "setFontSize":
            portName = call.argument("portName");
            double size = call.argument("size");
            printer = printerReceipts.get(portName);
            printer.receipt.setFontSize((float)size);
            result.success(null);;
            break;
          case "setCommand":
            portName = call.argument("portName");
            String command = call.argument("command");
            printer = printerReceipts.get(portName);
            printer.receipt.setCommand(command);
            result.success(null);;
            break;
          case "closeReceipt":
            portName = call.argument("portName");
            printer = printerReceipts.get(portName);
            printer.receipt.closeReceipt();
            result.success(null);;
            break;
          case "printReceipt":
            portName = call.argument("portName");
            double delay = call.argument("delay");
            int retry = call.argument("retry");
            printer = printerReceipts.get(portName);
            if (printer != null) {
              printer.printer.printReceipt(printer.receipt, (float) delay, retry, status -> {
                HashMap<String, Object> hash = new HashMap<>();
                hash.put("offline", status.offline);
                hash.put("error", status.error);
                hash.put("outOfPaper", status.outOfPaper);
                hash.put("paperJam", status.paperJam);
                hash.put("printed", status.printed);
                result.success(hash);
              });
            }
            break;
          case "connect":
            portName = call.argument("portName");
            barcodeReader = new BarcodeReader(portName,activity);

            barcodeReader.setBarcodeScanner(new BarcodeReader.BarcodeCallback() {
              @Override
              public void onBarcodeRead(String code) {
                 HashMap<String, String> barcode = new HashMap<>();
                barcode.put("code",code);
                channel.invokeMethod("onBarcodeRead",barcode);
              }
            });
            result.success(null);;
            break;
          case "disconnect":
            portName = call.argument("portName");
            barcodeReader.disconnect();
            barcodeReader = null;
            result.success(null);;
          default:
            result.notImplemented();
            break;
        }
      }
    });

  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  @Override
  public void onDetachedFromActivity(){

  }
  @Override
  public void onReattachedToActivityForConfigChanges(ActivityPluginBinding binding){

  }

  @Override
  public void onAttachedToActivity(ActivityPluginBinding binding){
    activity = binding.getActivity();
  }


  @Override
  public void onDetachedFromActivityForConfigChanges(){
  }
}
